### PR TITLE
[Kotlin Pre-Migration] Compat 

### PR DESCRIPTION
**Pull request by codingtosh**
[Kotlin Pre-Migration] Compat 
package com.ichi2.compat

Add NonNull/Nullable annotations to the java file before converting it to Kotlin in a subsequent PR. 
Reason: Complex call heirarchy, minimse the chances of breakage post conversion

### CODE REVIEW

**Review by david-allison-1 — CHANGES_REQUESTED** [review]
This is a large class hierarchy - could the `Nullable/NonNull` attributes be added in a commit before this goes in, I believe this will reduce the chance of breakage

**Comment by codingtosh** [comment]
@david-allison-1 Sure, so I should add the annotations in the java file according to the nullability we got in the conversion(kotlin file), yes?

**Comment by david-allison-1** [comment]
> @david-allison-1 Sure, so I should add the annotations in the java file according to the nullability we got in the conversion(kotlin file), yes?

Yep! Sorry for the late reply

**Comment by mikehardy** [comment]
semi-dormant, converting to draft for PR filtering

**Comment by david-allison** [comment]
Still has a conflict

**Comment by mikehardy** [comment]
Took another conflict while waiting 😩 - the big conflict is easy - moveDirectory and copyDirectory are totally done now, so just delete, but the other one looks a little more subtle, those lines were changed by #10578 and I'm not sure exactly what the effect is here

**Comment by david-allison** [comment]
> Took another conflict while waiting 😩 - the big conflict is easy - moveDirectory and copyDirectory are totally done now, so just delete, but the other one looks a little more subtle, those lines were changed by #10578 and I'm not sure exactly what the effect is here

Almost certainly a change from `CompatHelper.getCompat` to `CompatHelper.compat`

**Comment by codingtosh** [comment]
@david-allison I set the nullability as per the kotlin output of the auto-conversion, PTAL.

> > @david-allison-1 Sure, so I should add the annotations in the java file according to the nullability we got in the conversion(kotlin file), yes?
> 
> Yep! Sorry for the late reply

**Comment by david-allison** [comment]
It feels incorrect. Context shouldn't be nullable:

https://github.com/ankidroid/Anki-Android/blob/e670d4ca3551b8d70b711eb6242dba5b9a126076/AnkiDroid/src/main/java/com/ichi2/compat/CompatV26.java#L73-L76

**Comment by codingtosh** [comment]
@david-allison I've set the nullity according to my best judgement, disregarding the auto-conversion kotlin output(it was way off). PTAL.

**Review by Arthur-Milchior — COMMENTED** [review]
Hi,

Thanks a lot for this work.
I admit it's quite hard to follow, because a lot of, admittedly small, things occurs at the same time, and each would require a lot of work to check. 
Worse, I am writing done two remarks on unrelated topics, and that will delay the whole PR, while a lot of lines are clearly valid.

So, if you agree to do this extra work, I'd love:
* that mMinutes and mHours is separated, so we can discuss it without blocking Compat (unless you agree that this change is not necessary)
* to do one PR by method changed. This will ensure that each function that is obvious can get merged immediately, and we can concentrate separately on each function that would require discussion. Even better would be if you can explain in the commit message how you made this decision. E.g. because it's always called with non null value/because it sends its value to methods that requires non null value
* that you add the annotation to the inheriting method. The idea is that this will add the inheriting method to the diff, and so I can very easily check that the methods that take NonNull parameter actually assume the parameter is NonNull

If this is too much work for you, I accept to do it. Or if other maintainers believes it's too much precaution, I'm also fine with accepting all NonNull annotations (after all, at worst, we'll get type error. That's less a problem than if we accepted Null values). However, I'd still love explanations on my two remarks

**Comment by codingtosh** [comment]
> that mMinutes and mHours is separated, so we can discuss it without blocking Compat (unless you agree that this change is not necessary)

@Arthur-Milchior Thanks for the detailed review, splitting this into multiple PR sounds great. 

I will defer the changes of the methods taking TimePicker as the param for a separate PR and will elucidate my reasoning and findings in the new PR and also brief it in the commit message.

> Why "nullable"?
I guess you may have chosen this because NotificationChannel takes String, which may be Null in theory. However, a quick look at the documentation seems to indicate it expects two strings. A quick look at our code seems to indicate we never call it with null values. So NonNull seems appropriate

Yup, my reasoning was based on the possibility of the null string. Your suggested change for setupNotificationChannel() sounds reasonable, I will include it in the next patchset I upload. 

> that you add the annotation to the inheriting method. The idea is that this will add the inheriting method to the diff, and so I can very easily check that the methods that take NonNull parameter actually assume the parameter is NonNull

I like this idea too, I had mentioned this in the @KotlinCleanup too and planned to revisit it after it got merged. Including it in this PR itself sounds much better, thanks for the suggestion.

Just needed a little clarification, since the rest of the nullable/non-null attributions seem reasonable to you, should I also write down the reasoning for each and every one of the methods(and their parameters) within the commit message of this PR?  

This also seems to have taken a conflict, will also remove that and upload the suggested changes tomorrow, off to 😴 🛌 for today.

**Comment by Arthur-Milchior** [comment]
thanks @mikehardy for adding tags I forgot once again to add.
Thanks @codingtosh for accepting those comments, as I know I ask a lot on top of what I imagine was already not a really exciting work.

If that's not too much work, I'd appreciate each function to have its own commit. The message itself may simply be "the implementation already assumes its non null" (which will be easy to see if you annotate inheriting method in the same commit), or "looked at all call, and it's always called with NonNullable value".
I understand it may be boring work. But it seems to easy to later get a NullPointerException or a failed compilation once moved to Kotlin with non-nullable type, and a single of those error would probably cost us globally more time than the time I'm asking right now to convince that your annotations are the right one.

**Comment by codingtosh** [comment]
@Arthur-Milchior @david-allison It keeps attracting conflicts when it just sits, we should merge this sooner than later. Please take a look.

**Review by david-allison — APPROVED** [review]

**Review by mikehardy — APPROVED** [review]
You have been quiet determined working to get this in shape a few times, thanks a bunch for that. Running it through local rebase/re-integration test and if it passes I'll merge it right now before bitrot gets it again!

**Comment by mikehardy** [comment]
note: I'm going to squash it but the commit-by-commit changes were key to me being able to review it quickly, that was much appreciated

**Comment by mikehardy** [comment]
made it!
